src/layouts/hexbin.js

```diff
@@ -0,0 +1,112 @@
+const w0 = Math.sin(Math.PI / 3);
+
+function hbin(I, X, Y, r) {
+  const dx = r * 2 * w0;
+  const dy = r * 1.5;
+  const keys = new Map();
+  return groups(I, i => {
+    let px = X[i] / dx;
+    let py = Y[i] / dy;
+    if (isNaN(px) || isNaN(py)) return;
+    let pj = Math.round(py),
+        pi = Math.round(px = px - (pj & 1) / 2),
+        py1 = py - pj;
+    if (Math.abs(py1) * 3 > 1) {
+      let px1 = px - pi,
+        pi2 = pi + (px < pi ? -1 : 1) / 2,
+        pj2 = pj + (py < pj ? -1 : 1),
+        px2 = px - pi2,
+        py2 = py - pj2;
+      if (px1 * px1 + py1 * py1 > px2 * px2 + py2 * py2) pi = pi2 + (pj & 1 ? 1 : -1) / 2, pj = pj2;
+    }
+    const key = `${pi}|${pj}`;
+    keys.set(key, [pi, pj]);
+    return key;
+  })
+  .filter(([p]) => p)
+  .map(([p, bin]) => {
+    const [pi, pj] = keys.get(p);
+    bin.x = (pi + (pj & 1) / 2) * dx;
+    bin.y = pj * dy;
+    return bin;
+  });
+}
+
+// Allow hexbin options to be specified as part of outputs; merge them into options.
+function mergeOptions({radius = 10, ...outputs}, options) {
+  return [outputs, {radius, ...options}];
+}
+
+function hexbinLayout(radius, outputs, options) {
+  // we defer to Plot.bin’s reducers, but some of them are not supported
+  for (const reduce of Object.values(outputs)) {
+    if (typeof reduce === "string"
+    && !reduce.match(/^(first|last|count|distinct|sum|deviation|min|min-index|max|max-index|mean|median|variance|mode|proportion|proportion-facet)$/i))
+      throw new Error(`invalid reduce ${reduce}`);
+  }
+  outputs = maybeOutputs(outputs, options);
+  const rescales = {
+    r: {scale: "r", options: {range: [0, radius * w0]}},
+    fill: {scale: "color"},
+    stroke: {scale: "color"},
+    fillOpacity: {scale: "opacity"},
+    strokeOpacity: {scale: "opacity"},
+    symbol: {scale: "symbol"}
+  };
+  const {x, y} = options;
+  if (x == null) throw new Error("missing channel: x");
+  if (y == null) throw new Error("missing channel: y");
+  return layout({...defaults, ...options}, function(index, scales, {x: X, y: Y}) {
+    const values = {x: [], y: [], r: []};
+    const channels = [];
+    const newIndex = [];
+    for (const o of outputs) {
+      o.initialize(this.data);
+      o.scope("data", index);
+    }
+    let n = 0;
+    for (const I of index) {
+      const facetIndex = [];
+      newIndex.push(facetIndex);
+      const bins = hbin(I, X, Y, radius);
+      for (const o of outputs) {
+        o.scope("facet", I);
+        for (const bin of bins) o.reduce(bin);
+      }
+      for (const bin of bins) {
+        values.x[n] = bin.x;
+        values.y[n] = bin.y;
+        facetIndex.push(n++);
+      }
+    }
+    for (const o of outputs) {
+      if (o.name in rescales) {
+        const {scale, options} = rescales[o.name];
+        const value = o.output.transform();
+        channels.push([o.name, {scale, value, options}]);
+      } else {
+        values[o.name] = o.output.transform();
+      }
+    }
```

Comment:
This is using the “materialized” representation of a channel, similar to that returned by the [Channel constructor](https://github.com/observablehq/plot/blob/2b59461f5c05107922da3ba0a602fb3faa271a2f/src/channel.js#L7-L16), rather than the channel “descriptors” that are used e.g. by the Mark constructor and *mark*.channels. Channel descriptors are materialized inside of the *mark*.initialize, given data.

I wonder if it would be more appropriate to use the “channel descriptor” representation here, rather than having layouts return materialized channels, so as to avoid introducing a new public signature for channel objects.

This is the sort of thing that makes me wish we used TypeScript, since we’d be forced to be explicit about the shapes of these objects. We’ll get there eventually I guess. 😄